src/migrations.rs

```diff
@@ -38,9 +48,117 @@ impl Contract {
+            access_control,
+            authors,
+        });
+    }
+
+    fn unsafe_insert_old_post_authors(start: u64, end: u64) -> StateVersion {
+        let mut contract: Contract = env::state_read().unwrap();
+        let total = contract.posts.len();
+        let end = min(total, end);
+        for i in start..end {
+            let versioned_post = contract.posts.get(i);
+            if let Some(versioned_post) = versioned_post {
+                let post: Post = versioned_post.into();
+                let mut author_posts =
+                    contract.authors.get(&post.author_id).unwrap_or_else(|| HashSet::new());
+                author_posts.insert(post.id);
+                contract.authors.insert(&post.author_id, &author_posts);
+            }
+        }
+        env::state_write(&contract);
+        StateVersion::V3 { done: end == total, migrated_count: end }
+    }
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Debug)]
+enum StateVersion {
+    V1,
+    V2,
+    V3 { done: bool, migrated_count: u64 },
+}
+
+const VERSION_KEY: &[u8] = b"VERSION";
+
+fn state_version_read() -> StateVersion {
+    env::storage_read(VERSION_KEY)
+        .map(|data| {
+            StateVersion::try_from_slice(&data).expect("Cannot deserialize the contract state.")
+        })
+        .unwrap_or(StateVersion::V2) // StateVersion is introduced in production contract with V2 State.
+}
+
+fn state_version_write(version: &StateVersion) {
+    let data = version.try_to_vec().expect("Cannot serialize the contract state.");
+    env::storage_write(VERSION_KEY, &data);
+    near_sdk::log!("Migrated to version: {:?}", version);
+}
+
+#[near_bindgen]
+impl Contract {
+    pub fn unsafe_self_upgrade() {
+        near_sdk::assert_self();
+
+        let contract = env::input().expect("No contract code is attached in input");
+        Promise::new(env::current_account_id()).deploy_contract(contract);
+    }
+
+    pub fn unsafe_migrate() {
+        near_sdk::assert_self();
+        let current_version = state_version_read();
+        near_sdk::log!("Migrating from version: {:?}", current_version);
+        match current_version {
+            StateVersion::V1 => {
+                Contract::unsafe_add_acl();
+                state_version_write(&StateVersion::V2);
+            }
+            StateVersion::V2 => {
+                Contract::unsafe_add_post_authors();
+                state_version_write(&StateVersion::V3 { done: false, migrated_count: 0 })
+            }
+            StateVersion::V3 { done: false, migrated_count } => {
+                let new_version =
+                    Contract::unsafe_insert_old_post_authors(migrated_count, migrated_count + 100);
+                state_version_write(&new_version);
+                if let StateVersion::V3 { done: true, migrated_count: _ } = new_version {
+                    near_sdk::log!("Migration finished.");
+                    env::value_return(&true.try_to_vec().unwrap());
```

Comment:
Is it going to serialize the response with borsh? Let's return a JSON string `"needs-migration"` or `"done"`, so it is explicit. A boolean is hard to extend and reason about.

Reply:
Will do! Yeah that will be easiser to handle on the CD.